xmtp_mls/src/groups/device_sync.rs

```diff
@@ -0,0 +1,596 @@
+        if let Ok((_msg, request)) = self.pending_sync_request(request.kind).await {
+            return Ok((request.request_id, request.pin_code));
+        }
+
+        // build the request
+        let request: DeviceSyncRequestProto = request.into();
+        let pin_code = request.pin_code.clone();
+        let request_id = request.request_id.clone();
+
+        let content = DeviceSyncContent::Request(request.clone());
+        let content_bytes = serde_json::to_vec(&content)?;
+
+        let _message_id = sync_group.prepare_message(&content_bytes, &conn, move |_time_ns| {
+            PlaintextEnvelope {
+                content: Some(Content::V2(V2 {
+                    message_type: Some(Request(request)),
+                    idempotency_key: new_request_id(),
+                })),
+            }
+        })?;
+
+        // publish the intent
+        if let Err(err) = sync_group.publish_intents(&conn.into()).await {
+            tracing::error!("error publishing sync group intents: {:?}", err);
+        }
+
+        Ok((request_id, pin_code))
+    }
+
+    async fn send_sync_reply(&self, contents: DeviceSyncReplyProto) -> Result<(), DeviceSyncError> {
+        // find the sync group
+        let conn = self.store().conn()?;
+        let sync_group = self.get_sync_group()?;
+
+        // sync the group
+        sync_group.sync().await?;
+
+        let (msg, _request) = self.pending_sync_request(contents.kind()).await?;
+
+        // add original sender to all groups on this device on the node
+        self.ensure_member_of_all_groups(&msg.sender_inbox_id)
+            .await?;
+
+        // the reply message
+        let (content_bytes, contents) = {
+            let content = DeviceSyncContent::Reply(contents);
+            let content_bytes = serde_json::to_vec(&content)?;
+            let DeviceSyncContent::Reply(contents) = content else {
+                unreachable!("This is a reply.");
+            };
+
+            (content_bytes, contents)
+        };
+
+        let _message_id = sync_group.prepare_message(&content_bytes, &conn, move |_time_ns| {
+            PlaintextEnvelope {
+                content: Some(Content::V2(V2 {
+                    idempotency_key: new_request_id(),
+                    message_type: Some(Reply(contents)),
+                })),
+            }
+        })?;
+
+        // publish the intent
+        if let Err(err) = sync_group.publish_messages().await {
+            tracing::error!("error publishing sync group intents: {:?}", err);
+        }
+        Ok(())
+    }
+
+    async fn pending_sync_request(
+        &self,
+        kind: DeviceSyncKind,
+    ) -> Result<(StoredGroupMessage, DeviceSyncRequestProto), DeviceSyncError> {
+        let sync_group = self.get_sync_group()?;
+
+        sync_group.sync().await?;
+
+        let messages = sync_group
+            .find_messages(&MsgQueryArgs::default().kind(GroupMessageKind::Application))?;
+
+        for msg in messages.into_iter().rev() {
```

Comment:
would it be worth adding a field to `group_messages` here, or maybe something else to make this lookup faster? @neekolas

Reply:
I'd definitely be down for this optimization later if it becomes an issue. Right now I don't think it'll be much of an issue since the design right now is to create a new sync group every time a device is added, which becomes the default sync group for all installations.